### PR TITLE
Exercise4

### PR DESCRIPTION
**Pull request by canyavuzkurt**

### CODE REVIEW

File: force-app/main/default/classes/OpenWeatherChecker.cls [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * Description: Invokes OpenWeather to receive weather condition
+ * *************************************************************************
+ * History:
+ * Date:                Modified By:                         Description:
+ * August 19, 2021      Can Yavuzkurt                        Created
+ * August 20, 2021      Can Yavuzkurt                        Updated endpoint url
+ */
+public with sharing class OpenWeatherChecker {
+    
+    private static String apiKey = 'cd5155cd5e4a5d3751e0e98a4374bcb4';
```
Comment (romanbychkoosf) [inline force-app/main/default/classes/OpenWeatherChecker.cls]: as it is constant syntax have to be
private static final String API_KEY = '';

File: force-app/main/default/classes/OpenWeatherChecker.cls [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * @author Can Yavuzkurt
+     * @date August 19, 2021
+     * @description Retrieves the current weather condition of the given city
+     *              from openweathermap.org
+     * @name checkCurrentCondition
+     * @param cityName String city to retrieve the current condition of
+    */
+    public static CurrentWeatherWrapper checkCurrentCondition(String cityName) {
+        if (cityName == null) {
```
Comment (romanbychkoosf) [inline force-app/main/default/classes/OpenWeatherChecker.cls]: it is better to check with
if(String.isBlank(cityName)){

also please check the difference between String.isBlank and String.isEmpty

File: force-app/main/default/classes/OpenWeatherChecker.cls [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            
+            throw new HandledException('cityName param cannot be null');
+        }
+
+        Http http = new Http();
+        HttpRequest request = new HttpRequest();
+        request.setEndpoint('https://api.openweathermap.org/data/2.5/weather?q=' + cityName + '&APPID=' + apiKey);
+        request.setMethod('GET');
+        HttpResponse response = http.send(request);
```
Comment (romanbychkoosf) [inline force-app/main/default/classes/OpenWeatherChecker.cls]: it has to be in try/catch

File: force-app/main/default/classes/OpenWeatherChecker.cls [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    */
+    public static CurrentWeatherWrapper checkCurrentCondition(String cityName) {
+        if (cityName == null) {
+            
+            throw new HandledException('cityName param cannot be null');
+        }
+
+        Http http = new Http();
+        HttpRequest request = new HttpRequest();
+        request.setEndpoint('https://api.openweathermap.org/data/2.5/weather?q=' + cityName + '&APPID=' + apiKey);
```
Comment (romanbychkoosf) [inline force-app/main/default/classes/OpenWeatherChecker.cls]: please move endpoint into constants

File: force-app/main/default/classes/OpenWeatherChecker.cls [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        request.setMethod('GET');
+        HttpResponse response = http.send(request);
+        // If the request is successful, parse the JSON response.
+        if(response.getStatusCode() == 200) {
```
Comment (romanbychkoosf) [inline force-app/main/default/classes/OpenWeatherChecker.cls]: it is bettwe to check with ranges, like if >= 200 AND < 300

File: force-app/main/default/classes/OpenWeatherChecker.cls [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            // Deserialize the JSON string into collections of primitive data types.
+            return CurrentWeatherWrapper.parse(response.getBody());
+        }
+        else if (response.getStatusCode() == 404) {
```
Comment (romanbychkoosf) [inline force-app/main/default/classes/OpenWeatherChecker.cls]: here also it is bettwe to use range

File: force-app/main/default/classes/OpenWeatherCheckerTest.cls [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ * Created By:  Can Yavuzkurt
+ * *************************************************************************
+ * Description:  Unit test for OpenWeatherChecker.cls
+ * *************************************************************************
+ * History:
+ * Date:                Modified By:                         Description:
+ * August 19, 2021      Can Yavuzkurt                        Created
+ */
+@IsTest
+private class OpenWeatherCheckerTest {
```
Comment (romanbychkoosf) [inline force-app/main/default/classes/OpenWeatherCheckerTest.cls:14]: Please create 2 "mock" classes, 1 for success, 2nd for bad request

**Review by romanbychkoosf — CHANGES_REQUESTED** [review]

File: force-app/main/default/classes/OpenWeatherChecker.cls (hunk 6 above)
Comment (canyavuzkurt) [reply force-app/main/default/classes/OpenWeatherChecker.cls]: In this one I am throwing an exception specifically for 404. Rest of the codes goes to else block. Should I use a range anyway?

Comment (romanbychkoosf) [reply force-app/main/default/classes/OpenWeatherChecker.cls]: Yes, I think for Bad request (from 400 to 499) it is better to use separate if. And all the rest (500 codes) are in else

File: force-app/main/default/classes/OpenWeatherChecker.cls (hunk 1 above)
Comment (canyavuzkurt) [reply force-app/main/default/classes/OpenWeatherChecker.cls]: Updated

File: force-app/main/default/classes/OpenWeatherChecker.cls (hunk 2 above)
Comment (canyavuzkurt) [reply force-app/main/default/classes/OpenWeatherChecker.cls]: Updated

File: force-app/main/default/classes/OpenWeatherChecker.cls (hunk 4 above)
Comment (canyavuzkurt) [reply force-app/main/default/classes/OpenWeatherChecker.cls]: Updated

File: force-app/main/default/classes/OpenWeatherChecker.cls (hunk 5 above)
Comment (canyavuzkurt) [reply force-app/main/default/classes/OpenWeatherChecker.cls]: Updated

File: force-app/main/default/classes/OpenWeatherChecker.cls (hunk 6 above)
Comment (canyavuzkurt) [reply force-app/main/default/classes/OpenWeatherChecker.cls]: Alright got you. Updated

File: force-app/main/default/classes/OpenWeatherCheckerTest.cls (hunk 7 above)
Comment (canyavuzkurt) [reply force-app/main/default/classes/OpenWeatherCheckerTest.cls:14]: Updated

**Review by romanbychkoosf — APPROVED** [review]

**Pull request merged**